tests/two_d_primitive_tests.cpp

```diff
@@ -0,0 +1,106 @@
+#include "two_d_primitive_tests.h"
+
+#include <SDL_image.h>
+#include <pbkit/pbkit.h>
+
+#include "../nxdk_ext.h"
+#include "../pbkit_ext.h"
+#include "../test_host.h"
+#include "debug_output.h"
+#include "vertex_buffer.h"
+
+// From pbkit.c
+#define MAXRAM 0x03FFAFFF
+
+// See pb_init in pbkit.c, where the channel contexts are set up.
+// SUBCH_4 == GR_CLASS_62, which contains the NV10 2D surface commands.
+#define SUBCH_CLASS_62 SUBCH_4
+
+// 5C is the class for NV04_RENDER_SOLID_LIN (nv32.h) / NV04_SOLID_LINE (nv_objects.h)
+static constexpr uint32_t SUBCH_CLASS_5C = kNextSubchannel;
+// 42 is the class for NV04_CONTEXT_SURFACES_2D (nv32.h) & (nv_objects.h)
+static constexpr uint32_t SUBCH_CLASS_42 = SUBCH_CLASS_5C + 1;
+
+static constexpr TwoDPrimitiveTests::TwoDPrimitiveTest kTests[] = {
+    {0x000000, 400, 200, 400, 400},
+    {0xFF0000, 400, 200, 400, 400},
+    {0x00FF00, 400, 200, 400, 400},
+    {0x0000FF, 400, 200, 400, 400},
+    {0xFFFFFF, 0, 0, 799, 599},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 400, 0, 400, 599},
+    {0xFFFFFF, 400, 300, 401, 300},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 999, 300, 0, 300},
+};
+
+TwoDPrimitiveTests::TwoDPrimitiveTests(TestHost& host, std::string output_dir)
+    : TestSuite(host, std::move(output_dir)) {
+  for (auto test : kTests) {
+    std::string name = MakeTestName(test);
+
+    auto test_method = [this, test]() { this->Test(test); };
+    tests_[name] = test_method;
+  }
+}
+
+void TwoDPrimitiveTests::Initialize() {
+  SetDefaultTextureFormat();
+
+  auto channel = kNextContextChannel;
+
+  pb_create_gr_ctx(channel++, NV04_SOLID_LINE, &solid_lin_ctx_);
+  pb_bind_channel(&solid_lin_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_5C, &solid_lin_ctx_);
+
+  pb_create_gr_ctx(channel++, NV04_CONTEXT_SURFACES_2D, &surface_destination_ctx_);
+  pb_bind_channel(&surface_destination_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_42, &surface_destination_ctx_);
+
+  // set up to not have shader
+  // no vertex shader (transform shader)..
+  // no fragment or pixel shader (color from triangle)..
+  host_.SetShaderProgram(nullptr);
+}
+
+void TwoDPrimitiveTests::Test(const TwoDPrimitiveTest& test) {
+  host_.PrepareDraw(0xFF440011); //beta + RRGGBB
```

Comment:
This is actually `AARRGGBB`, beta is a blend factor whereas alpha is the actual transparency value.

Reply:
fixed